src/plugins/expressions/common/expression_functions/specs/math.ts

```diff
@@ -0,0 +1,163 @@
+
+function pivotObjectArray<
+  RowType extends { [key: string]: any },
+  ReturnColumns extends string | number | symbol = keyof RowType
+>(rows: RowType[], columns?: string[]): Record<string, ReturnColumns[]> {
+  const columnNames = columns || Object.keys(rows[0]);
+  if (!columnNames.every(isString)) {
+    throw new Error('Columns should be an array of strings');
+  }
+
+  const columnValues = map(columnNames, (name) => map(rows, name));
+  return zipObject(columnNames, columnValues);
+}
+
+export const errors = {
+  emptyExpression: () =>
+    new Error(
+      i18n.translate('expressions.functions.math.emptyExpressionErrorMessage', {
+        defaultMessage: 'Empty expression',
+      })
+    ),
+  tooManyResults: () =>
+    new Error(
+      i18n.translate('expressions.functions.math.tooManyResultsErrorMessage', {
+        defaultMessage:
+          'Expressions must return a single number. Try wrapping your expression in {mean} or {sum}',
+        values: {
+          mean: 'mean()',
+          sum: 'sum()',
+        },
+      })
+    ),
+  executionFailed: () =>
+    new Error(
+      i18n.translate('expressions.functions.math.executionFailedErrorMessage', {
+        defaultMessage: 'Failed to execute math expression. Check your column names',
+      })
+    ),
+  emptyDatatable: () =>
+    new Error(
+      i18n.translate('expressions.functions.math.emptyDatatableErrorMessage', {
+        defaultMessage: 'Empty datatable',
+      })
+    ),
+};
+
+const fallbackValue = {
+  null: null,
+  zero: 0,
+  false: false,
+} as const;
+
+export const math: ExpressionFunctionDefinition<
+  'math',
+  MathInput,
+  MathArguments,
+  boolean | number | null
+> = {
+  name: 'math',
+  type: undefined,
+  inputTypes: ['number', 'datatable'],
+  help: i18n.translate('expressions.functions.mathHelpText', {
+    defaultMessage:
+      'Interprets a {TINYMATH} math expression using a {TYPE_NUMBER} or {DATATABLE} as {CONTEXT}. ' +
+      'The {DATATABLE} columns are available by their column name. ' +
+      'If the {CONTEXT} is a number it is available as {value}.',
+    values: {
+      TINYMATH,
+      CONTEXT: '_context_',
+      DATATABLE: '`datatable`',
+      value: '`value`',
+      TYPE_NUMBER: '`number`',
+    },
+  }),
+  args: {
+    expression: {
+      aliases: ['_'],
+      types: ['string'],
+      help: i18n.translate('expressions.functions.math.args.expressionHelpText', {
+        defaultMessage: 'An evaluated {TINYMATH} expression. See {TINYMATH_URL}.',
+        values: {
+          TINYMATH,
+          TINYMATH_URL,
+        },
+      }),
+    },
+    onError: {
+      types: ['string'],
+      help: i18n.translate('expressions.functions.math.args.onErrorHelpText', {
+        defaultMessage:
+          "In case the {TINYMATH} evaluation fails or returns NaN, the return value is specified by onError. When `'throw'`, it will throw an exception, terminating expression execution (default).",
```

Comment:
This should specify valid values instead of just being of type string (to allow auto-complete)

Reply:
I tried that as first option, but could not find a way to map a string union to a specific type for expressions. All the examples I check in the documentation have the `string` type assigned and documenting the valid values on the documentation side.